### PR TITLE
fix(processing): Trim frames hard in stacktraces

### PR DESCRIPTION
**Pull request by mitsuhiko**

### CODE REVIEW

File: src/sentry/interfaces/stacktrace.py [hunk 1]
```diff
@@ -218,6 +219,8 @@ def slim_frame_data(frames, frame_allowance=settings.SENTRY_MAX_STACKTRACE_FRAME
         else:
             system_frames.append(frame)
 
+    del frames[hard_frame_limit - 1:-1]
```
Comment (alex-hofsteede) [inline src/sentry/interfaces/stacktrace.py]: `:-1` is incorrect I think. It leaves the last frame in place. Perhaps that is intended though?

Comment (mitsuhiko) [reply src/sentry/interfaces/stacktrace.py]: That was intentional though not sure if great

File: tests/sentry/interfaces/test_stacktrace.py [hunk 2]
```diff
@@ -639,6 +639,25 @@ def test_collapse_recursion(self):
...
 
+    def test_frame_hard_limit(self):
+        interface = Stacktrace.to_python(
+            {
+                'frames': [
+                    {
+                        'filename': 'Application.java',
+                        'function': 'main',
+                        'lineno': i,
+                    } for i in range(1, 252)
```
Comment (alex-hofsteede) [inline tests/sentry/interfaces/test_stacktrace.py]: hardcoded to be equal to the current limit from settings, which is not the best

File: src/sentry/interfaces/stacktrace.py [hunk 3]
```diff
@@ -203,7 +204,8 @@ def remove_module_outliers(module, platform=None):
     return module
 
 
-def slim_frame_data(frames, frame_allowance=settings.SENTRY_MAX_STACKTRACE_FRAMES):
+def slim_frame_data(frames, frame_allowance=settings.SENTRY_MAX_STACKTRACE_FRAMES,
+                    hard_frame_limit=settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT):
```
Comment (EvanPurkhiser) [inline src/sentry/interfaces/stacktrace.py]: This parameter isn't being used.

File: src/sentry/interfaces/stacktrace.py [hunk 4]
```diff
@@ -691,9 +693,15 @@ def to_python(cls, data, slim_frames=True, raw=False):
         if not is_valid:
             raise InterfaceValidationError("Invalid stack frame data.")
 
+        # trim down the frame list for excessive things
+        frameiter = data['frames']
+        if len(data['frames']) > settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT:
+            frameiter = chain(
+                islice(data['frames'], settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT - 1), (data['frames'][-1],))
```
Comment (EvanPurkhiser) [inline src/sentry/interfaces/stacktrace.py]: Why does this need the `chain`? Can't this just be

```py
if len(data['frames']) > settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT:
  frameiter = islice(data['frames'], settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT)
```

maybe an off by one somewhere in there though

File: tests/sentry/interfaces/test_stacktrace.py (hunk 2 above)
Comment (EvanPurkhiser) [reply tests/sentry/interfaces/test_stacktrace.py]: Can we just pull in `settings.SENTRY_STACKTRACE_FRAMES_HARD_LIMIT` and add some number, that's a little better at least

**Review by evanpurkhiser — APPROVED** [review]
great work @mitsuhiko @alex-hofsteede

**Pull request merged**